File: src/main/java/com/race/codeDash/mapper/CodeMapper.java
package com.race.codeDash.mapper;

import com.race.codeDash.dto.CodeDto;
import com.race.codeDash.infrastructure.TupleLineCode;
import org.springframework.stereotype.Component;

import java.util.LinkedHashMap;
import java.util.LinkedList;


@Component
public class CodeMapper {

	public TupleLineCode<LinkedList<Integer>, LinkedList<CodeDto>> parseCodeSnippet(String code) {
		LinkedList<CodeDto> codeData = new LinkedList<>();
		LinkedList<Integer> lineNumbers = new LinkedList<>();
		Integer lineCount = 1;
		code = code.replaceAll("\r\n", "\n");

		while (!code.isEmpty() && (code.charAt(0) == '\n' || code.charAt(0) == '\r')) {
			code = code.substring(1);
		}

		for (int i = 0; i < code.length(); i++) {
			String character = code.substring(i, i + 1);
			if (character.equals("\n") || character.equals("\r")) {
				character = "⏎";
				lineNumbers.add(lineCount++);
			}
			else if (character.equals(" ")){
				character = "space";
			}
			if (i == code.length() - 1 && !character.equals("⏎")) lineNumbers.add(lineCount++);

			codeData.add(new CodeDto(String.valueOf(i), character));
		}
		return new TupleLineCode<>(lineNumbers,codeData);
	}


}

File: codeDashPhp/public/model/mapper/CodeMapper.php
<?php

require_once ROOT_PATH . "src/TupleLineCode.php";
require_once ROOT_PATH . "model/dto/CodeDto.php";

class CodeMapper {

    public function parseCodeSnippet($code) {
        $codeData = [];
        $lineNumbers = [];
        $lineCount = 1;
        $code = str_replace("\r\n", "\n", $code);

        while (!empty($code) && ($code[0] == '\n' || $code[0] == '\r')) {
            $code = substr($code, 1);
        }

        for ($i = 0; $i < strlen($code); $i++) {
            $character = substr($code, $i, 1);
            if ($character == "\n" || $character == "\r") {
                $character = "⏎";
                $lineNumbers[] = $lineCount++;
            } elseif ($character == " ") {
                $character = "space";
            }
            if ($i == strlen($code) - 1 && $character != "⏎") {
                $lineNumbers[] = $lineCount++;
            }

            $codeData[] = new CodeDto((string)$i, $character);
        }

        return new TupleLineCode($lineNumbers, $codeData);
    }
}